### PR TITLE
Sanitize charm origin

### PR DESCRIPTION
**Pull request by SimonRichardson**
The following aims to sanitize the charm origin before it leaves the API
server. Although not required, it ensures that we prevent the leaking of
the platform types out to the client.

The idea is that if we send "arch=amd64" and they send back "arch=all",
then we can replace "all" with "amd64" and the client is none the wiser.

The one difference is that we will recalculate the OS from the series if
the series in the platform is labeled as "all".

This will once and for all kill off the idea of all leaking out.

## QA steps

```sh
$ juju bootstrap lxd test
$ juju deploy wiki-simple
$ juju deploy ubuntu
```

### CODE REVIEW

File: apiserver/facades/client/charms/repositories.go [hunk 1]
```diff
@@ -151,6 +170,42 @@ func (c *chRepo) findBundleDownloadURL(curl *charm.URL, origin corecharm.Origin)
...
+		os, err := series.GetOSFromSeries(requested.Series)
+		if err != nil {
+			return received, errors.Trace(err)
+		}
+		result.OS = os.String()
+	}
+	return result, nil
+}
+
+func sanitizeCoreCharmOrigin(received, requested corecharm.Origin) (corecharm.Origin, error) {
```
Comment (SimonRichardson) [inline apiserver/facades/client/charms/repositories.go:199]: Fun with types :face_with_head_bandage:

File: apiserver/facades/client/charms/repositories.go [hunk 2]
```diff
@@ -151,6 +170,42 @@ func (c *chRepo) findBundleDownloadURL(curl *charm.URL, origin corecharm.Origin)
...
+		result.OS = requested.OS
+	}
+	if received.Series == "all" {
+		result.Series = requested.Series
+
+		os, err := series.GetOSFromSeries(requested.Series)
+		if err != nil {
+			return received, errors.Trace(err)
+		}
+		result.OS = os.String()
```
Comment (hmlanigan) [inline apiserver/facades/client/charms/repositories.go:194]: why overwrite OS if done at line 187? os must be lower case, and should be in the received, or the call would have failed.

File: apiserver/facades/client/charms/repositories.go [hunk 3]
```diff
@@ -151,6 +170,42 @@ func (c *chRepo) findBundleDownloadURL(curl *charm.URL, origin corecharm.Origin)
...
+	}
+	return result, nil
+}
+
+func sanitizeCoreCharmOrigin(received, requested corecharm.Origin) (corecharm.Origin, error) {
+	a := apicharm.CoreCharmOrigin(received)
+	b := apicharm.CoreCharmOrigin(received)
```
Comment (hmlanigan) [inline apiserver/facades/client/charms/repositories.go:201]: a & b are the same
s/received/requested/

**Review by hmlanigan — COMMENTED** [review]
Adding comments, will qa after they are resolved.

**Review by hmlanigan — APPROVED** [review]
qa good.

**Comment by SimonRichardson** [comment]
$$merge$$

**Pull request merged**